[SPARK-12165][SPARK-12189] Fix bugs in eviction of storage memory by execution

JoshRosen opened the pull request:
This patch fixes a bug in the eviction of storage memory by execution.
## The bug:

In general, execution should be able to evict storage memory when the total storage memory usage is greater than `maxMemory * spark.memory.storageFraction`. Due to a bug, however, Spark might wind up evicting no storage memory in certain cases where the storage memory usage was between `maxMemory * spark.memory.storageFraction` and `maxMemory`. For example, here is a regression test which illustrates the bug:

``` scala
    val maxMemory = 1000L
    val taskAttemptId = 0L
    val (mm, ms) = makeThings(maxMemory)
    // Since we used the default storage fraction (0.5), we should be able to allocate 500 bytes
    // of storage memory which are immune to eviction by execution memory pressure.

    // Acquire enough storage memory to exceed the storage region size
    assert(mm.acquireStorageMemory(dummyBlock, 750L, evictedBlocks))
    assertEvictBlocksToFreeSpaceNotCalled(ms)
    assert(mm.executionMemoryUsed === 0L)
    assert(mm.storageMemoryUsed === 750L)

    // At this point, storage is using 250 more bytes of memory than it is guaranteed, so execution
    // should be able to reclaim up to 250 bytes of storage memory.
    // Therefore, execution should now be able to require up to 500 bytes of memory:
    assert(mm.acquireExecutionMemory(500L, taskAttemptId, MemoryMode.ON_HEAP) === 500L) // <--- fails by only returning 250L
    assert(mm.storageMemoryUsed === 500L)
    assert(mm.executionMemoryUsed === 500L)
    assertEvictBlocksToFreeSpaceCalled(ms, 250L)
```

The problem relates to the control flow / interaction between `StorageMemoryPool.shrinkPoolToReclaimSpace()` and `MemoryStore.ensureFreeSpace()`. While trying to allocate the 500 bytes of execution memory, the `UnifiedMemoryManager` discovers that it will need to reclaim 250 bytes of memory from storage, so it calls `StorageMemoryPool.shrinkPoolToReclaimSpace(250L)`. This method, in turn, calls `MemoryStore.ensureFreeSpace(250L)`. However, `ensureFreeSpace()` first checks whether the requested space is less than `maxStorageMemory - storageMemoryUsed`, which will be true if there is any free execution memory because it turns out that `MemoryStore.maxStorageMemory = (maxMemory - onHeapExecutionMemoryPool.memoryUsed)` when the `UnifiedMemoryManager` is used.

The control flow here is somewhat confusing (it grew to be messy / confusing over time / as a result of the merging / refactoring of several components). In the pre-Spark 1.6 code, `ensureFreeSpace` was called directly by the `MemoryStore` itself, whereas in 1.6 it's involved in a confusing control flow where `MemoryStore` calls `MemoryManager.acquireStorageMemory`, which then calls back into `MemoryStore.ensureFreeSpace`, which, in turn, calls `MemoryManager.freeStorageMemory`.
## The solution:

The solution implemented in this patch is to remove the confusing circular control flow between `MemoryManager` and `MemoryStore`, making the storage memory acquisition process much more linear / straightforward. The key changes:
- Remove a layer of inheritance which made the memory manager code harder to understand (53841174760a24a0df3eb1562af1f33dbe340eb9).
- Move some bounds checks earlier in the call chain (13ba7ada77f87ef1ec362aec35c89a924e6987cb).
- Refactor `ensureFreeSpace()` so that the part which evicts blocks can be called independently from the part which checks whether there is enough free space to avoid eviction (7c68ca09cb1b12f157400866983f753ac863380e).
- Realize that this lets us remove a layer of overloads from `ensureFreeSpace` (eec4f6c87423d5e482b710e098486b3bbc4daf06).
- Realize that `ensureFreeSpace()` can simply be replaced with an `evictBlocksToFreeSpace()` method which is called [after we've already figured out](https://github.com/apache/spark/blob/2dc842aea82c8895125d46a00aa43dfb0d121de9/core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala#L88) how much memory needs to be reclaimed via eviction; (2dc842aea82c8895125d46a00aa43dfb0d121de9).

Along the way, I fixed some problems with the mocks in `MemoryManagerSuite`: the old mocks would [unconditionally](https://github.com/apache/spark/blob/80a824d36eec9d9a9f092ee1741453851218ec73/core/src/test/scala/org/apache/spark/memory/MemoryManagerSuite.scala#L84) report that a block had been evicted even if there was enough space in the storage pool such that eviction would be avoided.

I also fixed a problem where `StorageMemoryPool._memoryUsed` might become negative due to freed memory being double-counted when excution evicts storage. The problem was that `StorageMemoryPoolshrinkPoolToFreeSpace` would [decrement `_memoryUsed`](https://github.com/apache/spark/commit/7c68ca09cb1b12f157400866983f753ac863380e#diff-935c68a9803be144ed7bafdd2f756a0fL133) even though `StorageMemoryPool.freeMemory` had already decremented it as each evicted block was freed. See SPARK-12189 for details.

JoshRosen committed:
Reset ensureFreeSpaceCalled before each test.

This reduces coupling between failed tests.

JoshRosen committed:
Add MemoryStore.freeSpaceForExecution() method, which forces blocks to be dropped.

Previously, ensureFreeSpace() might end up not dropping blocks if the total
storage memory pool usage was less than the maximum possible storage pool usage.

JoshRosen committed:
Add regression test for storage eviction bug.

JoshRosen committed:
Remove a layer of confusing inheritance.

JoshRosen committed:
Put fail-fast for non-fitting blocks earlier in call chain.

JoshRosen committed:
Collapse ensureFreeSpace overloads

JoshRosen committed:
Replace ensureFreeSpace() with evictBlocksToFreeSpace().

JoshRosen committed:
Update JIRA in test case.

SparkQA commented:
**[Test build #47251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47251/consoleFull)** for PR 10170 at commit [`0eac7da`](https://github.com/apache/spark/commit/0eac7da041326cfbec2c1db2f279cb655744f90e).

SparkQA commented:
**[Test build #47251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47251/consoleFull)** for PR 10170 at commit [`0eac7da`](https://github.com/apache/spark/commit/0eac7da041326cfbec2c1db2f279cb655744f90e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `  public abstract static class PrefixComputer `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47251/
Test FAILed.

JoshRosen commented:
I think that the test failures in `BlockManagerSuite` are related to the semantics of `maxNumBytesToFree` in `StaticMemoryManager.acquireUnrollMemory`.

[Previously](https://github.com/apache/spark/pull/10170/files#diff-935c68a9803be144ed7bafdd2f756a0fL87), it looks like `maxNumBytesToFree` was the amount of memory that we'd try to free in the `memoryStore.ensureFreeSpace()` call but due to the bug fixed in this patch I believe that we'd end up in the case where `maxNumBytesToFree` is less than the amount of free storage memory, so `ensureFreeSpace()` wouldn't evict anything even though we actually need `numBytesToAcquire > maxNumBytesToFree` bytes of memory.

After the fixes implemented here, we'll first claim as much free storage memory as possible, then subtract that from our memory goal and request the remaining memory via spilling. As a result, we are more prone to evict, which might be throwing off the original test case (it's a little tricky to say due to size estimation; I'll try to see if I can decouple that via mocking in order to make the test easier to reason about).

One minor question of semantics: up until now (and still) it looks like `maxNumBytesToFree` has been "the maximum amount of memory to attempt to free up via spilling" and not "an upper bound on the amount of memory that will be spilled in response to a request". I believe that this might be the right interpretation, since the latter interpretation would mean that large cached blocks could never be evicted by small unroll requests.

@andrewor14, is the original idea behind `spark.storage.unrollFraction` that it places an upper limit on the amount of total storage memory that can be used for unrolling? If that's the case, we can simplify `StorageMemoryPool.acquireMemory` to only accept a single `numBytesToAcquire` and cap the size of requests at `min(spark.storage.unrollFraction * maxStorageMemory - currentUnrollMemory, requestedUnrollMemory)`.

andrewor14 commented on core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala:
```diff
@@ -73,27 +73,31 @@ class StorageMemoryPool(lock: Object) extends MemoryPool(lock) with Logging {
+    if (numBytesToAcquire > memoryFree && maxNumBytesToFree > 0) {
+      val additionalMemoryRequired = numBytesToAcquire - memoryFree
+      memoryStore.evictBlocksToFreeSpace(
+        Some(blockId), Math.min(maxNumBytesToFree, additionalMemoryRequired), evictedBlocks)
```
can we use `math.min` like we do in other places

andrewor14 replied:
actually, to improve readability a little:
```
val val numBytesToFree = math.min(maxNumBytesToFree, additionalMemoryRequired)
memoryStore.evictBlocksToFreeSpace(Some(blockId), numBytesToFree, evictedBlocks)
```

andrewor14 commented on core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala:
```diff
@@ -128,9 +132,11 @@ class StorageMemoryPool(lock: Object) extends MemoryPool(lock) with Logging {
       val evictedBlocks = new ArrayBuffer[(BlockId, BlockStatus)]
-      memoryStore.ensureFreeSpace(spaceToFree - spaceFreedByReleasingUnusedMemory, evictedBlocks)
+      memoryStore.evictBlocksToFreeSpace(
+        None, spaceToFree - spaceFreedByReleasingUnusedMemory, evictedBlocks)
```
maybe a more readable way to write this is:
```
val spaceFreedByReleasingUnusedMemory = math.min(...)
decrementPoolSize(spaceFreedByReleasingUnusedMemory)
val remainingSpaceToFree = spaceToFree - spaceFreedByReleasingUnusedMemory
if (remainingSpaceToFree > 0) {
  // If reclaiming free memory did not ...
} else {
  spaceFreedByReleasingUnusedMemory
}
```

andrewor14 commented on core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:
```diff
@@ -100,7 +100,7 @@ private[spark] class UnifiedMemoryManager private[memory] (
         // For now, we only support on-heap caching of data, so we do not need to interact with
         // the storage pool when allocating off-heap memory. This will change in the future, though.
-        super.acquireExecutionMemory(numBytes, taskAttemptId, memoryMode)
+        offHeapExecutionMemoryPool.acquireMemory(numBytes, taskAttemptId)
```
+1 to this change. It's much less brittle!

andrewor14 commented on core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:
```diff
@@ -110,6 +110,12 @@ private[spark] class UnifiedMemoryManager private[memory] (
+    if (numBytes > maxMemory) {
+      // Fail fast if the block simply won't fit
+      logInfo(s"Will not store $blockId as the required space ($numBytes bytes) exceeds our " +
+        s"memory limit ($maxMemory bytes)")
```
this check is actually duplicated now that you've removed it from `ensureFreeSpace`, but I can see that it stops us from even resizing the memory pools early so it's OK to keep here

andrewor14 replied:
should this check be `numBytes > maxStorageMemory`? I don't think it makes much of a difference in practice but that would seem more correct

JoshRosen replied:
Yep, good catch: here in UMM, `maxStorageMemory` is `maxMemory - onHeapExecutionMemoryPool.memoryUsed`, reflecting the fact that storage cannot evict execution.

andrewor14 commented on core/src/main/scala/org/apache/spark/memory/StaticMemoryManager.scala:
```diff
@@ -53,6 +53,20 @@ private[spark] class StaticMemoryManager(
+      blockId: BlockId,
+      numBytes: Long,
+      evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = synchronized {
+    if (numBytes > storageMemoryPool.poolSize) {
```
if you make the other one (in UMM) `numBytes > maxStorageMemory` then it would make sense to do the same here, so it's more consistent (small readability improvement)

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/MemoryManagerSuite.scala:
```diff
@@ -36,105 +37,98 @@ import org.apache.spark.storage.{BlockId, BlockStatus, MemoryStore, StorageLevel
+   * [[assertEvictBlocksToFreeSpaceCalled]].
    */
-  private val ensureFreeSpaceCalled = new AtomicLong(DEFAULT_ENSURE_FREE_SPACE_CALLED)
+  private val evictBlocksToFreeSpaceCalled = new AtomicLong(0)
```
should this start at the `DEFAULT`?

andrewor14 replied:
I see, you do this in `beforeEach` anyway

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/MemoryManagerSuite.scala:
```diff
@@ -36,105 +37,98 @@ import org.apache.spark.storage.{BlockId, BlockStatus, MemoryStore, StorageLevel
-            (null, BlockStatus(StorageLevel.MEMORY_ONLY, numBytes, 0L, 0L)))
+            (null, BlockStatus(StorageLevel.MEMORY_ONLY, numBytesToFree, 0L, 0L)))
+          evictedBlocks.append(
+            (null, BlockStatus(StorageLevel.MEMORY_ONLY, numBytesToFree, 0L, 0L)))
```
as mentioned offline, this will likely add the same block to `evictedBlocks` twice

JoshRosen replied:
If you just add it to the one in the invocation args, then you won't be able to assert that execution memory requests resulted in the eviction of blocks, since the execution request doesn't pass in the `evictedBlocks` from the suite.

A simple fix is to only add if `!(args.last eq evictedBlocks)`.

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/StaticMemoryManagerSuite.scala:
```diff
@@ -85,33 +82,33 @@ class StaticMemoryManagerSuite extends MemoryManagerSuite {
     assert(mm.acquireStorageMemory(dummyBlock, 1L, evictedBlocks))
-    assertEnsureFreeSpaceCalled(ms, 1L)
+    assertEvictBlocksToFreeSpaceCalled(ms, 1L)
     assert(mm.storageMemoryUsed === 1000L)
```
It took me a while to understand why this one is still `1000L`. The situation is as follows:

1. Put a 1000 bytes block
2. Put a 1 byte block, this should have evicted the first block
3. Storage memory used is still 1000 bytes somehow (I expected 1 byte)

Because of the way the test is set up, however, as of step (2) we actually treat the 1000-byte block inserted in step (1) as 1000 1-byte blocks. When we call `evictBlocksToFreeSpace` here we actually evicted a 1-byte block to put another 1-byte block, so the storage memory usage remains at `1000L`.

Since this deviates from actual block manager behavior I think this line warrants a two-line comment, maybe something like:
```
// Note: We evicted 1 byte to put another 1-byte block in, so the storage memory used remains
// at 1000 bytes. This is different from real behavior, where the 1-byte block would have evicted
// the 1000-byte block entirely. This is set up differently so we can write finer-grained tests.
```

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/StaticMemoryManagerSuite.scala:
```diff
@@ -159,13 +156,13 @@ class StaticMemoryManagerSuite extends MemoryManagerSuite {
     assert(mm.acquireUnrollMemory(dummyBlock, 500L, evictedBlocks))
     // `spark.storage.unrollFraction` is 0.4, so the max unroll space is 400 bytes.
     // Since we already occupy 60 bytes, we will try to ensure only 400 - 60 = 340 bytes.
```
outdated comment

andrewor14 replied:
actually, this whole test should be rewritten to trigger the case where unroll evicts part of existing storage. I have some changes here https://github.com/andrewor14/spark/commit/78cd2be5efd14dcad4fcd648f6815d741c09694d that are passing tests. I'll submit a PR against your branch shortly.

Commit:
Rewrite unroll test after changes in eviction

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -78,33 +75,37 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
-    assertEnsureFreeSpaceCalled(ms, 1000L)
+    assertEvictBlocksToFreeSpaceCalled(ms, 110L)
     assert(mm.storageMemoryUsed === 1000L)
+    assert(evictedBlocks.nonEmpty)
```
OK to keep, but I don't think there's a lot to gain from asserting this

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -78,33 +75,37 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
     assert(mm.acquireStorageMemory(dummyBlock, 1L, evictedBlocks))
-    assertEnsureFreeSpaceCalled(ms, 1L)
+    assertEvictBlocksToFreeSpaceCalled(ms, 1L)
     assert(mm.storageMemoryUsed === 1000L)
```
same here as in the `StaticMemoryManagerSuite` case, we should add a comment to explain why this is still 1000

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -117,25 +118,29 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
     assert(mm.storageMemoryUsed === 750L)
-    assertEnsureFreeSpaceNotCalled(ms)
+    assertEvictBlocksToFreeSpaceNotCalled(ms)
+    assert(evictedBlocks.isEmpty)
```
should this assert just go into the body of the previous one? i.e. `assertEvictBlocksToFreeSpaceNotCalled` internally also checks that `evictedBlocks` is empty. I notice we have these 2 lines in quite a few places.

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -117,25 +118,29 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
     assert(mm.executionMemoryUsed === 300L)
-    assertEnsureFreeSpaceCalled(ms, 50L)
+    assertEvictBlocksToFreeSpaceCalled(ms, 50L)
     assert(mm.executionMemoryUsed === 300L)
```
duplicated assert? This should check `mm.storageMemoryUsed === 700L` instead

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -143,7 +148,25 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
+    assert(mm.executionMemoryUsed === 0L)
+    assert(mm.storageMemoryUsed === 750L)
+    // Should now be able to require up to 500 bytes of memory
+    assert(mm.acquireExecutionMemory(500L, taskAttemptId, MemoryMode.ON_HEAP) === 500L)
```
Should we briefly explain SPARK-12165 since it's not at all obvious? Also I would change the numbers a little so it's easier to explain. Maybe something like:
```
...
...
assert(mm.storageMemoryUsed === 700L)
// SPARK-12165: previously, MemoryStore would not evict anything because it would
// mistakenly think that the 300 bytes of free space is still available even after
// using it to expand the execution pool. Consequently, no blocks were dropped
// and the following call used to grant only 300 bytes to execution.
assert(mm.acquireExecutionMemory(500L, taskAttemptId, MemoryMode.ON_HEAP) === 500L)
assertEvictBlocksToFreeSpaceCalled(ms, 200L)
```

andrewor14 replied:
also, I think the test name should be something like
```
requesting execution memory less than free memory should evict storage (SPARK-12165)
```
The problem isn't that "storage memory is below than max memory". Even in the existing test "eviction evicts storage", storage memory (750) is below the max memory (1000).

andrewor14 replied:
By the way I included these changes as part of my PR: https://github.com/JoshRosen/spark/pull/8

JoshRosen replied:
> The problem isn't that "storage memory is below than max memory". Even in the existing test "eviction evicts storage", storage memory (750) is below the max memory (1000).

It's true that this was a little inprecise, but I think that "execution memory requests smaller than free memory should evict storage" is also slightly inprecise because this essentially sounds like a synonym for "execution evicts storage when there is not enough free memory." I guess a slightly more accurate description of SPARK-12165 would be something like "execution evicts storage if free memory is insufficient and free memory > 0", since not _all_ situations where there is insufficient free memory will trigger the first part of SPARK-12165. This precision doesn't really matter, though, so I'm happy to adopt your test name if you find it to be clearer.

andrewor14 commented on core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala:
```diff
@@ -73,27 +73,31 @@ class StorageMemoryPool(lock: Object) extends MemoryPool(lock) with Logging {
-      metrics.updatedBlocks = Some(lastUpdatedBlocks ++ evictedBlocks.toSeq)
+    if (numBytesToAcquire > memoryFree && maxNumBytesToFree > 0) {
+      val additionalMemoryRequired = numBytesToAcquire - memoryFree
+      memoryStore.evictBlocksToFreeSpace(
```
after looking at the test case I think we should add a check here to do this only if there's a chance that evicting can free us enough memory, i.e.
```
val additionalMemoryRequired = math.max(0, numBytesToAcquire - memoryFree)
val numBytesToFree = math.min(maxNumBytesToFree, additionalMemoryRequired)
// Only evict blocks if there is a chance that doing so will grant us what we want
if (numBytesToFree > 0 && numBytesToFree <= memoryUsed) {
  memoryStore.evictBlocksToFreeSpace(...)
  ...
}
```
If we do this then we can't have the case where we try to free 150B from evicting blocks when there are only 100B worth of blocks.

andrewor14 replied:
(I should add that I think the existing code is correct, but rewriting it would make the semantics clearer and the tests easier to follow)

JoshRosen replied:
`evictBlocksToFreeSpace` already has a form of this check internally. Even if we add the logic here, we can't drop the checks inside of `evictBlocksToFreeSpace` because when storage is evicting storage we need to handle the fact that an RDD cannot evict blocks belonging to the same RDD.

It wouldn't hurt to add the check here, I suppose, since it would save us the hassle of having to iterate over all of the blocks in the memory store.

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -154,32 +177,34 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
     assert(mm.executionMemoryUsed === 800L)
     assert(mm.storageMemoryUsed === 100L)
-    assertEnsureFreeSpaceCalled(ms, 250L)
+    assertEvictBlocksToFreeSpaceCalled(ms, 150L) // try to evict blocks ...
```
If you take my suggestion from above about adding the preemptive check, then this will not be called. (i.e. we know there's only 100B of blocks so don't bother trying to free 150B out of it)

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -154,32 +177,34 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
     assert(mm.executionMemoryUsed === 200L)
     assert(mm.storageMemoryUsed === 750L)
-    assertEnsureFreeSpaceCalled(ms, 750L)
+    assertEvictBlocksToFreeSpaceNotCalled(ms) // since there was 800 bytes free
```
was -> were

andrewor14 commented on core/src/test/scala/org/apache/spark/memory/UnifiedMemoryManagerSuite.scala:
```diff
@@ -154,32 +177,34 @@ class UnifiedMemoryManagerSuite extends MemoryManagerSuite with PrivateMethodTes
     assert(mm.executionMemoryUsed === 200L)
     assert(mm.storageMemoryUsed === 750L)
-    assertEnsureFreeSpaceCalled(ms, 850L)
+    assertEvictBlocksToFreeSpaceCalled(ms, 800L) // try to evict blocks...
```
same here, if you take the suggestion then this will not happen

JoshRosen replied:
Yep, this was addressed by changing `maxMemory` to `maxStorageMemory` in the check at the start of `acquireStorageMemory`.

andrewor14 commented:
@JoshRosen despite the number of comments I left I think this patch looks good. I did a close review to verify its correctness and that the two bugs were real issues. On the side I will investigate the test failures and hopefully get this merged soon.

Commit:
Clarify regression test for SPARK-12165

This commit also adds a regression test for SPARK-12189 through
the existing test "execution evicts storage" by adding an assert.

JoshRosen committed:
Merge pull request #8 from andrewor14/josh-unroll-tests

Rewrite some tests after changes in eviction

JoshRosen committed:
Incorporate Andrew's redability suggestions.

JoshRosen committed:
maxMemory -> maxStorageMemory in check for blocks that can never fit

SparkQA commented:
**[Test build #47378 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47378/consoleFull)** for PR 10170 at commit [`d182bdc`](https://github.com/apache/spark/commit/d182bdcf4649d2cf9d961177a8ea8764ba69e23e).

JoshRosen committed:
Avoid double-add to evictedBlocks.

SparkQA commented:
**[Test build #47380 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47380/consoleFull)** for PR 10170 at commit [`f6fb406`](https://github.com/apache/spark/commit/f6fb4066e52baa6bf6c3680db6cdc967330a110c).

JoshRosen committed:
was -> were

JoshRosen committed:
Add comments to explain potentially-confusing 1-byte block eviction test.

SparkQA commented:
**[Test build #47382 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47382/consoleFull)** for PR 10170 at commit [`e2090d1`](https://github.com/apache/spark/commit/e2090d1e11109b8bfd232d3a33d9acbf775e0872).

JoshRosen committed:
Push evictedBlocks checks into assertEvictBlocksToFreeSpaceNotCalled.

SparkQA commented:
**[Test build #47383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47383/consoleFull)** for PR 10170 at commit [`a43ed34`](https://github.com/apache/spark/commit/a43ed343f81dc2001718dcd1812ed85a1d849d72).

SparkQA commented:
**[Test build #47378 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47378/consoleFull)** for PR 10170 at commit [`d182bdc`](https://github.com/apache/spark/commit/d182bdcf4649d2cf9d961177a8ea8764ba69e23e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47378/
Test FAILed.

SparkQA commented:
**[Test build #47380 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47380/consoleFull)** for PR 10170 at commit [`f6fb406`](https://github.com/apache/spark/commit/f6fb4066e52baa6bf6c3680db6cdc967330a110c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47380/
Test FAILed.

JoshRosen commented:
Note to self RE: unroll fraction (just to be precise):

At a high-level, the StaticMemoryManager a request for unroll memory will be able to obtain more memory if:

- there is free storage memory, or
- `currentUnrollMemory / maxStorageMemory` < `spark.memory.unrollFraction`.

An unroll request can only evict blocks if  `finalUnrollMemory / maxStorageMemory` < `spark.memory.unrollFraction`, where `finalUnrollMemory` is the amount of unroll memory that would be used after the eviction was performed.

SparkQA commented:
**[Test build #47382 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47382/consoleFull)** for PR 10170 at commit [`e2090d1`](https://github.com/apache/spark/commit/e2090d1e11109b8bfd232d3a33d9acbf775e0872).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47382/
Test FAILed.

SparkQA commented:
**[Test build #47383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47383/consoleFull)** for PR 10170 at commit [`a43ed34`](https://github.com/apache/spark/commit/a43ed343f81dc2001718dcd1812ed85a1d849d72).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47383/
Test FAILed.

Commit:
Fix unroll semantics to match old behavior

Before this commit, unrolling would evict too many blocks,
resulting in test failures in BlockManagerSuite. The root cause
is that we used `maxUnrollMemory` as a cap for the extra amount
of memory to evict for unrolling, which is incorrect. Instead,
we should use it as a cap for the total amount of unroll memory
and calculate the amount of memory to evict from there.

The goal of this commit is to preserve the old behavior (in 1.5)
as much as possible. This can be seen from the fact that
BlockManagerSuite now passes without any modifications.

JoshRosen committed:
Merge pull request #9 from andrewor14/josh-unroll-tests

One way to fix the tests

andrewor14 commented:
LGTM, I'll merge this once tests pass.

JoshRosen commented:
Updated PR description to remove the following text from the end (which will be incorporated into separate PRs, most likely):


## TODOs

- [ ] Add stronger assertions or a dedicated regression test for the `_memoryUsed < 0` bug, which was uncovered while testing this patch and a related fix for SPARK-12155.
- [ ] See whether we can now remove the confusing `MemoryStore.maxMemory` method.

/cc @andrewor14 @zsxwing @yhuai @rxin

SparkQA commented:
**[Test build #47433 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47433/consoleFull)** for PR 10170 at commit [`7e9e191`](https://github.com/apache/spark/commit/7e9e191c76992813d9a0bdbb1de590776c4875a1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47433/
Test PASSed.

andrewor14 commented:
Merging into master and 1.6!!
